{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b6e2c126-75f0-4f44-8073-2013e21e869a",
   "metadata": {},
   "source": [
    "# Script GFF into FASTA\n",
    "Version GitHub 01\n",
    "\n",
    "Decompress all .gff.gz to .gff files in the current working folder  \n",
    "Transform all .GFF into .FNA  \n",
    "GZIP all .fna-files in the current folder into .fna.gz  \n",
    "\n",
    "########################################################  \n",
    "By Torben Kuehnast, torben.kuehnast@gmail.com, 2024\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc4c0138-e32e-4db7-92e6-c69ccafabd32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Decompress all .gff.gz to .gff files in the current working folder  \n",
    "import os\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "# Use the current working directory\n",
    "folder_path = os.getcwd()\n",
    "\n",
    "for filename in os.listdir(folder_path):\n",
    "    if filename.endswith(\".gff.gz\"):\n",
    "        gz_file_path = os.path.join(folder_path, filename)\n",
    "        with gzip.open(gz_file_path, 'rb') as f_in:\n",
    "            unzipped_file_path = os.path.join(folder_path, filename[:-3])  # Removing '.gz' from filename\n",
    "            with open(unzipped_file_path, 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "        print(f\"Decompressed {filename} to {filename[:-3]}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d93faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# Transform all .GFF (ALL in current folder!!!) into .FNA\n",
    "#\n",
    "# Some sequencing output is given in .gff format, containing mainly genome annotation,\n",
    "# but also sometimes the full genome at the end of the file.\n",
    "# This tool extracts the genome sequence from a XXX.gff file in FASTA format and\n",
    "# saves it as XXX.fna. Everything after ##FASTA is copied to the new file.\n",
    "\n",
    "####################################\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "\n",
    "files = glob.glob('*.gff')\n",
    "for file in files:\n",
    "    print(file)\n",
    "    filepath_gff = str(file)\n",
    "    togg = False\n",
    "    print(\"toggle:\", togg)\n",
    "\n",
    "    pre_ou = filepath_gff\n",
    "    print(\"pre_ou:\", pre_ou)\n",
    "\n",
    "    # Get the file's name, and split .gff away from basename to get the file's name only.\n",
    "    ou = os.path.basename(pre_ou).split('.gff', 1)[0]\n",
    "    print('ou:', ou)\n",
    "\n",
    "    # Load the gff file\n",
    "    fi = open(filepath_gff)\n",
    "    print(\"fi:\", fi)\n",
    "\n",
    "    # Create a new file and call it basename + .fna ending in write mode\n",
    "    w = open(ou+\".fna\",'w')\n",
    "    print(\"w\", w)\n",
    "\n",
    "    for line in fi:\n",
    "        if \"##FASTA\" in line:\n",
    "            togg = True\n",
    "        elif togg:\n",
    "            w.write(line)\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "    print(\"last line:\", line)\n",
    "    print(\"togg (should be true):\", togg)\n",
    "\n",
    "    fi.close()\n",
    "    w.close()\n",
    "\n",
    "    print(\"File was created as \"+ou+\".fna\")\n",
    "    print(\"---- Endloop of \"+filepath_gff+\" ----\")\n",
    "print(\"Program finished.\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed282ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "####################################\n",
    "# GZIP all .fna-files in the current folder into .fna.gz\n",
    "#\n",
    "# Some sequencing output is given in .gff format, containing mainly genome annotation,\n",
    "# but also sometimes the full genome at the end of the file.\n",
    "# This tool extracts the genome sequence from a XXX.gff file in FASTA format and\n",
    "# saves it as XXX.fna. Everything after ##FASTA is copied to the new file.\n",
    "#\n",
    "####################################import os\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "folder_path = os.getcwd()\n",
    "\n",
    "for filename in os.listdir(folder_path):\n",
    "    if filename.endswith(\".fna\"):\n",
    "        file_path = os.path.join(folder_path, filename)\n",
    "        with open(file_path, 'rb') as f_in:\n",
    "            gz_file_path = f\"{file_path}.gz\"\n",
    "            with gzip.open(gz_file_path, 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "        print(f\"Gzipped {filename} to {filename}.gz\")\n",
    "    print(\"-------------------\")\n",
    "print(\"Finished! G-zipping of all .fna files to .fna.gz completed in\", folder_path)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "621466e0-2291-4690-90ff-ce3fc11dcfb1",
   "metadata": {},
   "source": [
    "# Script GFF into FASTA\n",
    "Finished!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8766a571-de51-4af6-891a-e679705b2c3e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
